### PR TITLE
Feat[NV-1868]: Initial Queue Metric Monitoring 

### PR DESCRIPTION
**Pull request by Cliftonz**
### What change does this PR introduce?
This pr adds a metric queue and repeatable job to sample each queue in Novu every second to have the ability to later dynamically scale on the number of messages in the queue.

### CODE REVIEW

File: apps/api/package.json [hunk 1]
```diff
@@ -57,6 +58,7 @@
     "class-validator": "^0.14.0",
     "compression": "^1.7.4",
     "cross-env": "^7.0.3",
+    "d": "^1.0.1",
```
Comment (scopsy) [inline apps/api/package.json:61]: @Cliftonz what this is?

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      });
+    };
+  }
+
+  private async jobHasCompleted(job): Promise<void> {
+    try {
+      Logger.log('Metric job Completed');
+    } catch (error) {
+      Logger.error('Failed to set job as completed', error);
+    }
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:115]: In which scenario this should throw? 
```suggestion
      Logger.log('Metric job Completed');
```

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  private async jobHasFailed(job, error): Promise<void> {
+    try {
+      Logger.log('Metric job failed', error);
+    } catch (anotherError) {
+      Logger.error(`Failed to set job as failed ${anotherError}`);
+    }
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:123]:
```suggestion
      Logger.error('Metric job failed', error);
```

File: apps/worker/src/app/workflow/workflow.module.ts [hunk 4]
```diff
@@ -110,6 +116,17 @@ const SERVICES = [
     provide: TriggerQueueService,
     useClass: TriggerProcessorQueueService,
   },
+  {
+    provide: WsQueueService,
+    useClass: WsQueueService,
+  },
```
Comment (scopsy) [inline apps/worker/src/app/workflow/workflow.module.ts:122]: Why is this needed here?

File: packages/application-generic/src/services/bull-mq.service.ts [hunk 5]
```diff
@@ -64,8 +78,15 @@ export class BullmqService {
     const WorkerClass = !BullmqService.pro
       ? Worker
       : require('@taskforcesh/bullmq-pro').WorkerPro;
+
+    let internalOptions: WorkerOptions;
+    if (options) {
+      internalOptions = options;
+    }
+    internalOptions.metrics = { maxDataPoints: MetricsTime.ONE_MONTH };
```
Comment (scopsy) [inline packages/application-generic/src/services/bull-mq.service.ts:86]:
```suggestion
    let internalOptions: WorkerOptions = {};
    if (options) {
      internalOptions = options;
    }
    internalOptions.metrics = { maxDataPoints: MetricsTime.ONE_MONTH };
```

internalOptions.metrics might fail because internal options can be potentially undefined.

File: packages/application-generic/src/services/queue.service.ts [hunk 6]
```diff
@@ -22,10 +21,9 @@ export class QueueService<T = unknown> {
   public readonly bullMqService: BullmqService;
   public readonly DEFAULT_ATTEMPTS = 3;
 
-  constructor() {
+  constructor(public name = 'standard') {
```
Comment (scopsy) [inline packages/application-generic/src/services/queue.service.ts:24]: We shouldn't allow to override this property. Why is it needed? this can lead to some really bad bugs and back compatibility issues.

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+function sumOfSquares(values: number[]): number {
+  const meanVal = mean(values);
+
+  return sum(values.map((value) => (value - meanVal) ** 2));
+}
+
+@Injectable()
+export class MetricQueueService extends QueueService<Record<string, never>> {
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:22]: I'm not sure regarding this extension here from QueueService, because the QueueService previously served the job processing queue, and I'm afraid extending it like this can create a lot of problems. @LetItRock could you help with reviewing this one?

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      await this.jobHasCompleted(job);
+    });
+
+    this.bullMqService.worker.on('failed', async (job, error) => {
+      await this.jobHasFailed(job, error);
+    });
+
+    void this.addToQueue('metric-job', {}, '', {
+      repeat: {
+        immediately: true,
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:40]: Note that immediatly can cause some weired behaviour when a new image is deployed. Would suggest to keep this on the pattern only.

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      await this.bullMqService.queue.drain();
+      await this.bullMqService.worker.close();
+    }
+  }
+
+  private getWorkerOpts(): WorkerOptions {
+    return {
+      ...this.bullConfig,
+      lockDuration: 90000,
+      concurrency: 200,
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:58]: The concurreny here should be 1, and lockDuration much lower rate in my opinion

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      settings: {},
+    } as WorkerOptions;
+  }
+
+  private getWorkerProcessor() {
+    return async () => {
+      return await new Promise(async (resolve, reject) => {
+        for (const queueService of this.token_list) {
+          this.bullMqService.createQueue(queueService.name, {
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:67]: This creates a global instance on the parent class, and can lead to a lot of weird behaviour. We should create a new bullMq service instance for each Queue we have here.

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+          Logger.log(`${process.env.NOVU_MANAGED_SERVICE}: managed`);
+          Logger.log(`MetricQueueService/${queueService.name}/completed`, JSON.stringify(successMetric));
+          Logger.log(`MetricQueueService/${queueService.name}/completed`, JSON.stringify(failMetric));
+          if (process.env.NOVU_MANAGED_SERVICE === 'true') {
+            Logger.log(`MetricQueueService/${queueService.name}/completed`, JSON.stringify(successMetric));
+            Logger.log(`MetricQueueService/${queueService.name}/completed`, JSON.stringify(failMetric));
+          } else {
+            nr.recordMetric(`MetricQueueService/${queueService.name}/completed`, successMetric);
+            nr.recordMetric(`MetricQueueService/${queueService.name}/failed`, failMetric);
+          }
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:104]: I'm quite confused here, could you eloborate on why we are only logging it on the NOBU MANAGED SERVICe vs actually sending this to newrelic? we need the exact opposite here. Or Am I missing something here?

**Review by scopsy — COMMENTED** [review]
@LetItRock could you also help with a review here?

File: apps/worker/src/app/workflow/services/workflow-queue.service.ts [hunk 12]
```diff
@@ -43,6 +43,7 @@ export class WorkflowQueueService extends QueueService<IJobData> {
     @Inject(forwardRef(() => CreateExecutionDetails)) private createExecutionDetails: CreateExecutionDetails
   ) {
     super();
+    Logger.warn('Workflow queue service created');
```
Comment (LetItRock) [inline apps/worker/src/app/workflow/services/workflow-queue.service.ts:46]: leftover?

File: apps/worker/src/app/workflow/workflow.module.ts (hunk 4 above)
Comment (LetItRock) [reply apps/worker/src/app/workflow/workflow.module.ts:122]: The `WsQueueService` doesn't have the worker logic and yes it's to get the metrics

File: packages/application-generic/src/services/bull-mq.service.ts (hunk 5 above)
Comment (LetItRock) [reply packages/application-generic/src/services/bull-mq.service.ts:86]:
```suggestion
    const internalOptions: WorkerOptions = {
       ...options,
       metrics: { maxDataPoints: MetricsTime.ONE_MONTH },
    }
```

File: packages/application-generic/src/services/queue.service.ts (hunk 6 above)
Comment (LetItRock) [reply packages/application-generic/src/services/queue.service.ts:24]: explained in the first comment in `MetricQueueService` file

File: apps/worker/src/app/workflow/services/metric-queue.service.ts (hunk 7 above)
Comment (scopsy) [reply apps/worker/src/app/workflow/services/metric-queue.service.ts:22]: Thank you for the explanation @LetItRock this makes sense now, appreciate it!

File: apps/worker/package.json [hunk 13]
```diff
@@ -27,36 +27,39 @@
...
     "@sentry/tracing": "^7.40.0",
+    "@types/newrelic": "^9.13.0",
     "axios": "^1.3.3",
     "body-parser": "^1.20.0",
     "bull": "^4.4.0",
     "bullmq": "^3.10.2",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.0",
     "cross-env": "^7.0.3",
+    "d": "^1.0.1",
```
Comment (scopsy) [inline apps/worker/package.json:49]: @Cliftonz Another ping regarding this dependency, where is it used?

File: apps/worker/src/app/workflow/services/metric-queue.service.ts [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+            count: metrics.failed.count,
+            total: failNumber == 0 ? 0 : sum(metrics.failed.data),
+            min: failNumber == 0 ? 0 : min(metrics.failed.data),
+            max: failNumber == 0 ? 0 : max(metrics.failed.data),
+            sumOfSquares: failNumber == 0 ? 0 : sumOfSquares(metrics.failed.data),
+          };
+
+          if (process.env.NOVU_MANAGED_SERVICE === 'true') {
+            nr.recordMetric(`MetricQueueService/${queueService.name}/completed`, successMetric);
+            nr.recordMetric(`MetricQueueService/${queueService.name}/failed`, failMetric);
```
Comment (scopsy) [inline apps/worker/src/app/workflow/services/metric-queue.service.ts:117]: @Cliftonz maybe I'm missing something, but wasn't the most important thing for this is to understand how many jobs we have in flight and pending right now? The completed and failed is quite nice, but I think it lacks the important metrics of how many pending, and how many processing in parallel at the moment. 

Also would be great to know how many are "scheduled" to the future.

File: apps/worker/package.json (hunk 13 above)
Comment (Cliftonz) [reply apps/worker/package.json]: This should be removed now

File: apps/worker/src/app/workflow/services/metric-queue.service.ts (hunk 14 above)
Comment (Cliftonz) [reply apps/worker/src/app/workflow/services/metric-queue.service.ts:121]: You are correct, I have updated the code to including getting waiting, active, and delayed jobs.

File: apps/worker/src/app/workflow/usecases/send-message/send-message-chat.usecase.ts [hunk 15]
```diff
@@ -136,6 +136,8 @@ export class SendMessageChat extends SendMessageBase {
       await this.createExecutionDetails.execute(
         CreateExecutionDetailsCommand.create({
           ...CreateExecutionDetailsCommand.getDetailsFromJob(command.job),
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
```
Comment (ainouzgali) [inline apps/worker/src/app/workflow/usecases/send-message/send-message-chat.usecase.ts:140]: I believe this one is not needed

**Review by ainouzgali — APPROVED** [review]
